### PR TITLE
Thread safe map

### PR DESCRIPTION
**Pull request by evsamsonov**
Fix for the issue #393

### CODE REVIEW

**Comment by mrfoe7** [comment]
Please, resolve conflics

File: conn.go [hunk 1]
```diff
@@ -171,7 +209,8 @@ func (c *conn) serveError() {
 			}
 			if handler := c.namespace(errMsg.namespace); handler != nil {
 				if handler.onError != nil {
-					handler.onError(c.namespaces[errMsg.namespace], errMsg.err)
+					ns, _ := c.namespaces.Get(errMsg.namespace)
```
[the remark this answers is not in the document]
Comment (evsamsonov) [reply conn.go]: Fixed

**Comment by evsamsonov** [comment]
> Please, resolve conflics

Resolved

File: conn.go [hunk 2]
```diff
@@ -31,12 +31,50 @@ type writePacket struct {
...
+func newNamespaces() *namespaces {
+	return &namespaces{
+		namespaces: make(map[string]*namespaceConn),
+	}
+}
+
+func (n *namespaces) Get(ns string) (*namespaceConn, bool) {
+	n.mu.RLock()
+	defer n.mu.RUnlock()
+	namespace, ok := n.namespaces[ns]
```
Comment (mrfoe7) [inline conn.go:48]: please, join 48 and 49 rows

File: conn.go [hunk 3]
```diff
@@ -31,12 +31,50 @@ type writePacket struct {
 	data []interface{}
 }
 
+type namespaces struct {
+	mu sync.RWMutex
```
Comment (mrfoe7) [inline conn.go:35]: please, move mu under namespaces

File: server.go [hunk 4]
```diff
@@ -2,16 +2,41 @@ package socketio
...
 import (
 	"net/http"
+	"sync"
 
 	"github.com/gomodule/redigo/redis"
 	"github.com/googollee/go-socket.io/engineio"
 )
 
+type namespaceHandlers struct {
+	mu       sync.RWMutex
```
Comment (mrfoe7) [inline server.go:12]: please, move mu under `namespaceHandler`

File: conn.go (hunk 2 above)
Comment (evsamsonov) [reply conn.go:48]: Do you mean this?
```
return n.namespaces[ns]
```
It don't compile
> not enough arguments to return

File: conn.go (hunk 3 above)
Comment (evsamsonov) [reply conn.go]: Done

File: server.go (hunk 4 above)
Comment (evsamsonov) [reply server.go]: Done

**Pull request merged**